### PR TITLE
Parallel tweaks

### PR DESCRIPTION
**Pull request by Zoxc**
Various parallel tweaks which brings the compilation time for `syntex_syntax` from 12.23 to 11.84 with 6 threads in non-incremental debug mode.

r? @Mark-Simulacrum

### CODE REVIEW

File: src/librustc_interface/passes.rs [hunk 1]
```diff
@@ -803,7 +803,16 @@ fn analysis(tcx: TyCtxt<'_>, cnum: CrateNum) -> Result<()> {
             },
             {
                 tcx.stability_index(LOCAL_CRATE);
-
+            },
+            {
+                // Prefetch in case lints are emitted.
+                tcx.lint_levels(LOCAL_CRATE);
```
Comment (bjorn3) [inline src/librustc_interface/passes.rs]:
```suggestion
                tcx.ensure().lint_levels(LOCAL_CRATE);
```

File: src/librustc_interface/passes.rs [hunk 2]
```diff
@@ -783,13 +783,24 @@ fn analysis(tcx: TyCtxt<'_>, cnum: CrateNum) -> Result<()> {
                 sess.time("looking_for_derive_registrar", || proc_macro_decls::find(tcx));
             },
             {
+                tcx.get_lang_items(LOCAL_CRATE);
```
[the remark this answers is not in the document]
Comment (Zoxc) [reply src/librustc_interface/passes.rs]: No, this to prevent all threads from immediately blocking on `get_lang_items` in the parallel loop below.

**Comment by Zoxc** [comment]
This is now based on https://github.com/rust-lang/rust/pull/68171.

**Review by michaelwoerister — COMMENTED** [review]
Looks good to me. I posted some minor questions below. 

You can r=me once https://github.com/rust-lang/rust/pull/68171 has landed.

File: src/librustc_interface/passes.rs [hunk 3]
```diff
@@ -809,38 +845,45 @@ fn analysis(tcx: TyCtxt<'_>, cnum: CrateNum) -> Result<()> {
...
-
-    sess.time("dumping_chalk_like_clauses", || {
-        rustc_traits::lowering::dump_program_clauses(tcx);
-    });
-
-    sess.time("MIR_effect_checking", || {
-        for def_id in tcx.body_owners() {
-            mir::transform::check_unsafety::check_unsafety(tcx, def_id)
-        }
+    sess.time("misc_checking_3", || {
```
[the remark this answers is not in the document]
Comment (Zoxc) [reply src/librustc_interface/passes.rs]: I don't think the names are great, but I do find them useful.

File: src/librustc_interface/passes.rs [hunk 4]
```diff
@@ -785,13 +785,36 @@ fn analysis(tcx: TyCtxt<'_>, cnum: CrateNum) -> Result<()> {
...
                     tcx.ensure().check_mod_const_bodies(local_def_id);
                 });
+            },
+            {
+                // Prefetch in case lints are emitted.
+                tcx.lint_levels(LOCAL_CRATE);
+            },
+            {
+                // Prefetch in case something needs printing.
+                tcx.visible_parent_map(LOCAL_CRATE);
```
[the remark this answers is not in the document]
Comment (Zoxc) [reply src/librustc_interface/passes.rs]: I think we'll have to pick on a case by case basis here whether to use `ensure` or not.

File: src/librustc_interface/passes.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (michaelwoerister) [reply src/librustc_interface/passes.rs]: OK, let's keep them.

File: src/librustc_data_structures/sync.rs [hunk 5]
```diff
@@ -157,7 +175,11 @@ cfg_if! {
             where A: FnOnce() -> RA,
                   B: FnOnce() -> RB
         {
-            (oper_a(), oper_b())
+            let panic = Lock::new(None);
+            let a = catch(&panic, oper_a);
+            let b = catch(&panic, oper_b);
+            resume(panic);
+            (a.unwrap(), b.unwrap())
```
[the remark this answers is not in the document]
Comment (Zoxc) [reply src/librustc_data_structures/sync.rs]: Which is fine, we only use a single panic value in rustc (excluding bugs).

**Comment by JohnCSimon** [comment]
assigning back to author because of merge conflicts.

**Comment by nikomatsakis** [comment]
I believe this is blocked on #68171, so updating status accordingly.

**Comment by Dylan-DPC** [comment]
@Zoxc any updates on this?

**Comment by nikomatsakis** [comment]
Closing this pull request as Zoxc is stepping back from compiler development; see rust-lang/team#316.

**Pull request closed without merging**